File: js/tilings/hex-parallelogram-weave.js
/**
 * Generates a hex-parallelogram weave board as a graph.
 *
 * The tiling is built on a unit triangular lattice.  It combines side-1
 * equilateral triangles, side-2 regular hexagons, and skinny parallelograms
 * whose long sides are 4 unit segments and short sides are 1 unit segment.
 * The longer edges are subdivided at unit intervals so graph neighbors can be
 * matched across partial edges.
 */
export function generateHexParallelogramWeaveBoard(options) {
  const { cols, rows, tileSize: side, colorCount, rng } = options;
  const periodI = 10;
  const periodJ = 5;
  const targetWidth = Math.max(cols, 1) * 3.2 * side;
  const targetHeight = Math.max(rows, 1) * Math.sqrt(3) * side;
  const pad = 4;
  const cropMarginX = periodI * side / 2;
  const cropMarginY = periodJ * Math.sqrt(3) * side / 4;

  const motif = [
    { kind: 'hex', i: 0, j: 2 },
    { kind: 'parallelogram', i: 3, j: 0, longDir: [1, 0], shortDir: [-1, 1] },
    { kind: 'parallelogram', i: 4, j: 1, longDir: [0, 1], shortDir: [1, 0] },
    { kind: 'triangle-up', i: 3, j: 4 },
    { kind: 'parallelogram', i: 0, j: 1, longDir: [1, 0], shortDir: [0, 1] },
    { kind: 'parallelogram', i: 1, j: 2, longDir: [-1, 1], shortDir: [1, 0] },
    { kind: 'triangle-down', i: 0, j: 2 },
    { kind: 'parallelogram', i: 9, j: 0, longDir: [0, 1], shortDir: [1, -1] },
    { kind: 'parallelogram', i: 9, j: 3, longDir: [-1, 1], shortDir: [0, 1] },
    { kind: 'triangle-up', i: 8, j: 0 },
    { kind: 'triangle-down', i: 5, j: 1 },
    { kind: 'hex', i: 5, j: 1 }
  ];

  const rawTiles = [];
  const repeatMinY = -pad;
  const repeatMaxY = Math.ceil(targetHeight / (periodJ * Math.sqrt(3) * side / 2)) + pad;
  const shearPad = Math.ceil(Math.max(0, repeatMaxY * periodJ) / (2 * periodI)) + 2;
  const reverseShearPad = Math.ceil(Math.max(0, -repeatMinY * periodJ) / (2 * periodI)) + 2;
  const repeatMinX = -pad - shearPad;
  const repeatMaxX = Math.ceil(targetWidth / (periodI * side)) + pad + reverseShearPad;

  for (let ry = repeatMinY; ry <= repeatMaxY; ry++) {
    for (let rx = repeatMinX; rx <= repeatMaxX; rx++) {
      const offsetI = rx * periodI;
      const offsetJ = ry * periodJ;
      for (const tile of motif) {
        const i = tile.i + offsetI;
        const j = tile.j + offsetJ;
        if (tile.kind === 'hex') {
          addRawTile('hex', hexVertexKeys(i, j));
        } else if (tile.kind === 'parallelogram') {
          addRawTile('parallelogram', parallelogramVertexKeys(i, j, tile.longDir, tile.shortDir));
        } else if (tile.kind === 'triangle-up') {
          addRawTile('triangle-up', smallTriangleVertexKeys('U', i, j));
        } else {
          addRawTile('triangle-down', smallTriangleVertexKeys('D', i, j));
        }
      }
    }
  }

  const cropped = rawTiles
    .map(tile => ({ ...tile, centroid: averagePoint(tile.points) }))
    .filter(tile => (
      tile.centroid[0] >= -cropMarginX && tile.centroid[0] <= targetWidth + cropMarginX &&
      tile.centroid[1] >= -cropMarginY && tile.centroid[1] <= targetHeight + cropMarginY
    ))
    .sort((a, b) => {
      if (Math.abs(a.centroid[1] - b.centroid[1]) > 1e-6) return a.centroid[1] - b.centroid[1];
      return a.centroid[0] - b.centroid[0];
    });

  const tiles = cropped.map((tile, id) => ({
    id,
    type: tile.type,
    colorId: Math.floor(rng() * colorCount),
    ownerId: null,
    points: tile.points.map(point => [...point]),
    neighbors: []
  }));

  const neighborSets = tiles.map(() => new Set());
  const edgeToTiles = new Map();
  cropped.forEach((tile, tileId) => {
    for (let i = 0; i < tile.vertexKeys.length; i++) {
      const a = tile.vertexKeys[i];
      const b = tile.vertexKeys[(i + 1) % tile.vertexKeys.length];
      const key = canonicalPairKey(a, b);
      if (!edgeToTiles.has(key)) edgeToTiles.set(key, []);
      edgeToTiles.get(key).push(tileId);
    }
  });

  for (const tileIds of edgeToTiles.values()) {
    if (tileIds.length < 2) continue;
    for (let i = 0; i < tileIds.length; i++) {
      for (let j = i + 1; j < tileIds.length; j++) {
        const a = tileIds[i];
        const b = tileIds[j];
        if (a !== b) {
          neighborSets[a].add(b);
          neighborSets[b].add(a);
        }
      }
    }
  }

  tiles.forEach((tile, id) => {
    tile.neighbors = [...neighborSets[id]].sort((a, b) => a - b);
  });

  let bbox = computeBoundingBox(tiles);
  for (const tile of tiles) {
    tile.points = tile.points.map(([x, y]) => [x - bbox.minX, y - bbox.minY]);
  }
  bbox = computeBoundingBox(tiles);

  const startTileIds = uniqueIds([
    closestTileId(tiles, 0, 0),
    closestTileId(tiles, bbox.maxX, bbox.maxY),
    closestTileId(tiles, bbox.maxX, 0),
    closestTileId(tiles, 0, bbox.maxY)
  ]);

  return {
    version: 1,
    generator: 'hex-parallelogram-weave',
    width: bbox.maxX,
    height: bbox.maxY,
    cols,
    rows,
    tiles,
    startTileIds
  };

  function addRawTile(type, vertexKeys) {
    rawTiles.push({
      type,
      vertexKeys,
      points: vertexKeys.map(key => latticePointFromKey(key, side))
    });
  }
}

function hexVertexKeys(i, j) {
  return subdividedPolygonKeys([
    [i + 2, j],
    [i + 4, j],
    [i + 4, j + 2],
    [i + 2, j + 4],
    [i, j + 4],
    [i, j + 2]
  ]);
}

function parallelogramVertexKeys(i, j, longDir, shortDir) {
  const longI = longDir[0] * 4;
  const longJ = longDir[1] * 4;
  const shortI = shortDir[0];
  const shortJ = shortDir[1];
  return subdividedPolygonKeys([
    [i, j],
    [i + longI, j + longJ],
    [i + longI + shortI, j + longJ + shortJ],
    [i + shortI, j + shortJ]
  ]);
}

function smallTriangleVertexKeys(kind, i, j) {
  const vertices = kind === 'D'
    ? [[i, j], [i + 1, j], [i, j + 1]]
    : [[i + 1, j + 1], [i, j + 1], [i + 1, j]];
  return vertices.map(([vi, vj]) => latticeKey(vi, vj));
}

function subdividedPolygonKeys(corners) {
  const keys = [];
  for (let sideIdx = 0; sideIdx < corners.length; sideIdx++) {
    const [ai, aj] = corners[sideIdx];
    const [bi, bj] = corners[(sideIdx + 1) % corners.length];
    const steps = Math.max(Math.abs(bi - ai), Math.abs(bj - aj));
    const stepI = (bi - ai) / steps;
    const stepJ = (bj - aj) / steps;
    for (let step = 0; step < steps; step++) {
      keys.push(latticeKey(ai + step * stepI, aj + step * stepJ));
    }
  }
  return keys;
}

function latticeKey(i, j) {
  return `${i},${j}`;
}

function latticePointFromKey(key, side) {
  const [i, j] = key.split(',').map(Number);
  return [side * (i + j / 2), side * Math.sqrt(3) * j / 2];
}

function canonicalPairKey(a, b) {
  return a < b ? `${a}~${b}` : `${b}~${a}`;
}

function averagePoint(points) {
  const sum = points.reduce((acc, [x, y]) => [acc[0] + x, acc[1] + y], [0, 0]);
  return [sum[0] / points.length, sum[1] / points.length];
}

function computeBoundingBox(tiles) {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;

  for (const tile of tiles) {
    for (const [x, y] of tile.points) {
      minX = Math.min(minX, x);
      minY = Math.min(minY, y);
      maxX = Math.max(maxX, x);
      maxY = Math.max(maxY, y);
    }
  }

  return { minX, minY, maxX, maxY };
}

function closestTileId(tiles, tx, ty) {
  let bestId = tiles[0]?.id ?? 0;
  let bestDist = Infinity;
  for (const tile of tiles) {
    const [cx, cy] = averagePoint(tile.points);
    const dist = (cx - tx) ** 2 + (cy - ty) ** 2;
    if (dist < bestDist) {
      bestDist = dist;
      bestId = tile.id;
    }
  }
  return bestId;
}

function uniqueIds(ids) {
  return [...new Set(ids.filter(id => Number.isInteger(id)))];
}

File: go-flood/pkg/tilings/hex_parallelogram_weave.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
	"sort"
	"strings"
)

type hexParallelogramWeaveRawTile struct {
	TileType   string
	VertexKeys []string
	Points     []core.Point
	Centroid   core.Point
}

type hexParallelogramWeaveMotifTile struct {
	Kind     string
	I        int
	J        int
	LongDir  [2]int
	ShortDir [2]int
}

// GenerateHexParallelogramWeaveBoard generates a woven triangular-lattice
// tiling made from side-1 triangles, side-2 regular hexagons, and skinny
// side-4-by-side-1 parallelograms. Longer edges are subdivided into unit
// segments so neighboring tiles can be matched across partial edges.
func GenerateHexParallelogramWeaveBoard(options Options) core.Board {
	side := options.TileSize
	const periodI = 10
	const periodJ = 5
	targetWidth := math.Max(float64(options.Cols), 1) * 3.2 * side
	targetHeight := math.Max(float64(options.Rows), 1) * math.Sqrt(3) * side
	pad := 4
	cropMarginX := periodI * side / 2
	cropMarginY := periodJ * math.Sqrt(3) * side / 4

	motif := []hexParallelogramWeaveMotifTile{
		{Kind: "hex", I: 0, J: 2},
		{Kind: "parallelogram", I: 3, J: 0, LongDir: [2]int{1, 0}, ShortDir: [2]int{-1, 1}},
		{Kind: "parallelogram", I: 4, J: 1, LongDir: [2]int{0, 1}, ShortDir: [2]int{1, 0}},
		{Kind: "triangle-up", I: 3, J: 4},
		{Kind: "parallelogram", I: 0, J: 1, LongDir: [2]int{1, 0}, ShortDir: [2]int{0, 1}},
		{Kind: "parallelogram", I: 1, J: 2, LongDir: [2]int{-1, 1}, ShortDir: [2]int{1, 0}},
		{Kind: "triangle-down", I: 0, J: 2},
		{Kind: "parallelogram", I: 9, J: 0, LongDir: [2]int{0, 1}, ShortDir: [2]int{1, -1}},
		{Kind: "parallelogram", I: 9, J: 3, LongDir: [2]int{-1, 1}, ShortDir: [2]int{0, 1}},
		{Kind: "triangle-up", I: 8, J: 0},
		{Kind: "triangle-down", I: 5, J: 1},
		{Kind: "hex", I: 5, J: 1},
	}

	rawTiles := []hexParallelogramWeaveRawTile{}
	addRawTile := func(tileType string, vertexKeys []string) {
		points := make([]core.Point, len(vertexKeys))
		for i, key := range vertexKeys {
			points[i] = hexParallelogramWeaveLatticePointFromKey(key, side)
		}
		rawTiles = append(rawTiles, hexParallelogramWeaveRawTile{
			TileType:   tileType,
			VertexKeys: vertexKeys,
			Points:     points,
		})
	}

	repeatMinY := -pad
	repeatMaxY := int(math.Ceil(targetHeight/(periodJ*math.Sqrt(3)*side/2))) + pad
	shearPad := int(math.Ceil(math.Max(0, float64(repeatMaxY*periodJ))/(2*periodI))) + 2
	reverseShearPad := int(math.Ceil(math.Max(0, float64(-repeatMinY*periodJ))/(2*periodI))) + 2
	repeatMinX := -pad - shearPad
	repeatMaxX := int(math.Ceil(targetWidth/(periodI*side))) + pad + reverseShearPad

	for ry := repeatMinY; ry <= repeatMaxY; ry++ {
		for rx := repeatMinX; rx <= repeatMaxX; rx++ {
			offsetI := rx * periodI
			offsetJ := ry * periodJ
			for _, tile := range motif {
				i := tile.I + offsetI
				j := tile.J + offsetJ
				switch tile.Kind {
				case "hex":
					addRawTile("hex", hexParallelogramWeaveHexVertexKeys(i, j))
				case "parallelogram":
					addRawTile("parallelogram", hexParallelogramWeaveParallelogramVertexKeys(i, j, tile.LongDir, tile.ShortDir))
				case "triangle-up":
					addRawTile("triangle-up", hexParallelogramWeaveSmallTriangleVertexKeys("U", i, j))
				case "triangle-down":
					addRawTile("triangle-down", hexParallelogramWeaveSmallTriangleVertexKeys("D", i, j))
				}
			}
		}
	}

	cropped := []hexParallelogramWeaveRawTile{}
	for _, tile := range rawTiles {
		tile.Centroid = hexParallelogramWeaveAveragePoint(tile.Points)
		if tile.Centroid[0] >= -cropMarginX && tile.Centroid[0] <= targetWidth+cropMarginX && tile.Centroid[1] >= -cropMarginY && tile.Centroid[1] <= targetHeight+cropMarginY {
			cropped = append(cropped, tile)
		}
	}

	sort.Slice(cropped, func(i, j int) bool {
		if math.Abs(cropped[i].Centroid[1]-cropped[j].Centroid[1]) > 1e-6 {
			return cropped[i].Centroid[1] < cropped[j].Centroid[1]
		}
		return cropped[i].Centroid[0] < cropped[j].Centroid[0]
	})

	tiles := make([]core.Tile, len(cropped))
	for id, tile := range cropped {
		points := make([]core.Point, len(tile.Points))
		copy(points, tile.Points)
		tiles[id] = core.Tile{
			ID:        id,
			ColorID:   int(options.RNG() * float64(options.ColorCount)),
			OwnerID:   nil,
			Points:    points,
			Neighbors: []int{},
		}
	}

	neighborSets := make([]map[int]bool, len(tiles))
	for i := range neighborSets {
		neighborSets[i] = map[int]bool{}
	}

	edgeToTiles := map[string][]int{}
	for tileID, tile := range cropped {
		for i, a := range tile.VertexKeys {
			b := tile.VertexKeys[(i+1)%len(tile.VertexKeys)]
			key := hexParallelogramWeaveCanonicalPairKey(a, b)
			edgeToTiles[key] = append(edgeToTiles[key], tileID)
		}
	}

	for _, tileIDs := range edgeToTiles {
		if len(tileIDs) < 2 {
			continue
		}
		for i := 0; i < len(tileIDs); i++ {
			for j := i + 1; j < len(tileIDs); j++ {
				a := tileIDs[i]
				b := tileIDs[j]
				if a != b {
					neighborSets[a][b] = true
					neighborSets[b][a] = true
				}
			}
		}
	}

	for id := range tiles {
		neighbors := make([]int, 0, len(neighborSets[id]))
		for neighborID := range neighborSets[id] {
			neighbors = append(neighbors, neighborID)
		}
		sort.Ints(neighbors)
		tiles[id].Neighbors = neighbors
	}

	minX, minY, _, _ := hexParallelogramWeaveBoundingBox(tiles)
	for i := range tiles {
		for j, p := range tiles[i].Points {
			tiles[i].Points[j] = core.Point{p[0] - minX, p[1] - minY}
		}
	}

	_, _, width, height := hexParallelogramWeaveBoundingBox(tiles)
	startTileIds := hexParallelogramWeaveUniqueIds([]int{
		hexParallelogramWeaveClosestTileID(tiles, 0, 0),
		hexParallelogramWeaveClosestTileID(tiles, width, height),
		hexParallelogramWeaveClosestTileID(tiles, width, 0),
		hexParallelogramWeaveClosestTileID(tiles, 0, height),
	})

	return core.Board{
		Version:      1,
		Generator:    "hex-parallelogram-weave",
		Width:        width,
		Height:       height,
		Cols:         options.Cols,
		Rows:         options.Rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}

func hexParallelogramWeaveHexVertexKeys(i, j int) []string {
	return hexParallelogramWeaveSubdividedPolygonKeys([][2]int{
		{i + 2, j},
		{i + 4, j},
		{i + 4, j + 2},
		{i + 2, j + 4},
		{i, j + 4},
		{i, j + 2},
	})
}

func hexParallelogramWeaveParallelogramVertexKeys(i, j int, longDir, shortDir [2]int) []string {
	longI := longDir[0] * 4
	longJ := longDir[1] * 4
	shortI := shortDir[0]
	shortJ := shortDir[1]
	return hexParallelogramWeaveSubdividedPolygonKeys([][2]int{
		{i, j},
		{i + longI, j + longJ},
		{i + longI + shortI, j + longJ + shortJ},
		{i + shortI, j + shortJ},
	})
}

func hexParallelogramWeaveSmallTriangleVertexKeys(kind string, i, j int) []string {
	vertices := [][2]int{}
	if kind == "D" {
		vertices = [][2]int{{i, j}, {i + 1, j}, {i, j + 1}}
	} else {
		vertices = [][2]int{{i + 1, j + 1}, {i, j + 1}, {i + 1, j}}
	}
	keys := make([]string, len(vertices))
	for idx, vertex := range vertices {
		keys[idx] = hexParallelogramWeaveLatticeKey(vertex[0], vertex[1])
	}
	return keys
}

func hexParallelogramWeaveSubdividedPolygonKeys(corners [][2]int) []string {
	keys := []string{}
	for sideIdx := 0; sideIdx < len(corners); sideIdx++ {
		a := corners[sideIdx]
		b := corners[(sideIdx+1)%len(corners)]
		steps := int(math.Max(math.Abs(float64(b[0]-a[0])), math.Abs(float64(b[1]-a[1]))))
		stepI := (b[0] - a[0]) / steps
		stepJ := (b[1] - a[1]) / steps
		for step := 0; step < steps; step++ {
			keys = append(keys, hexParallelogramWeaveLatticeKey(a[0]+step*stepI, a[1]+step*stepJ))
		}
	}
	return keys
}

func hexParallelogramWeaveLatticeKey(i, j int) string {
	return fmt.Sprintf("%d,%d", i, j)
}

func hexParallelogramWeaveLatticePointFromKey(key string, side float64) core.Point {
	parts := strings.Split(key, ",")
	if len(parts) != 2 {
		return core.Point{0, 0}
	}
	var i, j int
	_, _ = fmt.Sscanf(parts[0], "%d", &i)
	_, _ = fmt.Sscanf(parts[1], "%d", &j)
	return core.Point{
		side * (float64(i) + float64(j)/2),
		side * math.Sqrt(3) * float64(j) / 2,
	}
}

func hexParallelogramWeaveCanonicalPairKey(a, b string) string {
	if a < b {
		return a + "~" + b
	}
	return b + "~" + a
}

func hexParallelogramWeaveAveragePoint(points []core.Point) core.Point {
	sumX := 0.0
	sumY := 0.0
	for _, point := range points {
		sumX += point[0]
		sumY += point[1]
	}
	return core.Point{sumX / float64(len(points)), sumY / float64(len(points))}
}

func hexParallelogramWeaveBoundingBox(tiles []core.Tile) (float64, float64, float64, float64) {
	minX := math.MaxFloat64
	minY := math.MaxFloat64
	maxX := -math.MaxFloat64
	maxY := -math.MaxFloat64
	for _, tile := range tiles {
		for _, p := range tile.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}
	return minX, minY, maxX, maxY
}

func hexParallelogramWeaveClosestTileID(tiles []core.Tile, tx, ty float64) int {
	bestID := -1
	bestDist := math.MaxFloat64
	for _, tile := range tiles {
		center := hexParallelogramWeaveAveragePoint(tile.Points)
		dist := math.Hypot(center[0]-tx, center[1]-ty)
		if dist < bestDist {
			bestDist = dist
			bestID = tile.ID
		}
	}
	return bestID
}

func hexParallelogramWeaveUniqueIds(ids []int) []int {
	seen := map[int]bool{}
	unique := []int{}
	for _, id := range ids {
		if id < 0 || seen[id] {
			continue
		}
		seen[id] = true
		unique = append(unique, id)
	}
	return unique
}
